src/transaction.js

```diff
@@ -308,10 +309,96 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
     txTmp.ins[inIndex].script = ourScript
   }
 
-  // serialize and hash
-  var buffer = Buffer.allocUnsafe(txTmp.__byteLength(false) + 4)
-  buffer.writeInt32LE(hashType, buffer.length - 4)
-  txTmp.__toBuffer(buffer, 0, false)
+  // BIP143 sighash activated in BitcoinCash via 0x40 bit
+  if (hashType & Transaction.SIGHASH_BITCOINCASHBIP143) {
+    
+    if (types.Null(inAmount)) {
+      throw new Error('Bitcoin Cash sighash requires value of input to be signed.')
+    }
+
+    // gather PrevOuts and Sequences for hashing
+    var prevOutsBuf = Buffer.allocUnsafe(txTmp.ins.length * 36)
+    var prevOutsBufOffset = 0
+    var inputSequencesBuf = Buffer.allocUnsafe(txTmp.ins.length * 4)
+    var inputSequencesBufOffset = 0
+    var currentPrevOut = Buffer.allocUnsafe(36)
+    txTmp.ins.forEach(function(input, i) {
+      // gather PrevOut as buffer
+      var prevOutBuf = Buffer.allocUnsafe(input.hash.length + 4)
+
+      input.hash.copy(prevOutBuf, 0)
+      prevOutBuf.writeUInt32LE(input.index, prevOutBuf.length - 4)
+
+      if (i === inIndex) prevOutBuf.copy(currentPrevOut, 0)
+
+      prevOutBuf.copy(prevOutsBuf, prevOutsBufOffset)
+      prevOutsBufOffset += 36
+
+      // gather Sequence as buffer
+      var sequenceBuf = Buffer.allocUnsafe(4)
+
+      sequenceBuf.writeUInt32LE(input.sequence, 0)
+
+      sequenceBuf.copy(inputSequencesBuf, inputSequencesBufOffset)
+      inputSequencesBufOffset += 4
+    })
+
+    // gather outputs for hashing
+    var outputsBufferArray = []
+    var outputsBuffersLength = 0
+    txTmp.outs.forEach(function(output) {
+      var outputBuffer = Buffer.allocUnsafe(output.script.length + 8)
+      bufferutils.writeUInt64LE(outputBuffer, output.value, 0)
+      output.script.copy(outputBuffer, 8)
+      outputsBufferArray.push(outputBuffer)
+      outputsBuffersLength += outputBuffer.length
+    })
+    var outputsBuf = Buffer.allocUnsafe(outputsBuffersLength)
+    var outputsBufOffset = 0
+    outputsBufferArray.forEach(function(outputBuffer) {
+      outputBuffer.copy(outputsBuf, outputsBufOffset)
+      outputsBufOffset += outputBuffer.length
+    })
+
+    // HASH256 to get the hashPrevouts, hashSequence, and hashOutputs
+    var hashPrevouts = bcrypto.hash256(prevOutsBuf)
+    var hashSequence = bcrypto.hash256(inputSequencesBuf)
+    var hashOutputs = bcrypto.hash256(outputsBuf)
```

Comment:
This all comes from BIP143 - could you maybe use the functions from that part instead of duplicating the code again here?